pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -137,23 +140,60 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, pus
-			// check with the name of our service (cluster names are in the format outbound|<port>||<hostname>
-			// so, we can check if the cluster names contains the service name, and determine if it is deleted by that
-			for _, n := range watched.ResourceNames {
-				_, _, svcHost, _ := model.ParseSubsetKey(n)
-				if svcHost == host.Name(key.Name) {
-					removedClusterNames = append(removedClusterNames, n)
-				}
+		if key.Kind == gvk.ServiceEntry {
+			service := push.ServiceForHostname(proxy, host.Name(key.Name))
+			// SidecarScope.ServiceForHostname will return nil if the proxy doesn't care about the service OR it was deleted.
+			// we can cross-reference with WatchedResources to figure out which services were deleted.
+			if service == nil {
+				// we assume a service was deleted, it doesn't hurt to have positives for removedResources when a resource included
+				// does not exist.
+				// WatchedResources.ResourceNames will contain the names of the clusters it is subscribed to. We can
+				// check with the name of our service (cluster names are in the format outbound|<port>||<hostname>
+				// so, we can check if the cluster names contains the service name, and determine if it is deleted by that
+				removedClusterNames = append(removedClusterNames, getRemovedClusterNames(watched.ResourceNames, nil, key.Name)...)
+			} else {
+				services = append(services, service)
 			}
 		} else {
-			services = append(services, service)
+			// destination rules
+			cfg := push.DestinationRuleByName(proxy, key.Name)
+			if cfg == nil {
+				// a destinationrule was deleted, get the populated destinationRule from PrevSidecarScope
+				prevCfg := push.PrevDestinationRuleByName(proxy, key.Name)
+				if prevCfg == nil {
+					// something went wrong, skip -- we shouldn't be nil here
+					break
+				}
+				dr := prevCfg.Spec.(*networking.DestinationRule)
+				removedClusterNames = append(removedClusterNames, getRemovedClusterNames(watched.ResourceNames, nil, dr.Host)...)
+			} else {
+				// destination exists, was updated
+				// generate cluster based on service associated with destinationrule
+				// hostName can be a wildcard, and can match to multiple services
+				dr := cfg.Spec.(*networking.DestinationRule)
+				matchedSvcs := push.ServicesForHostname(proxy, host.Name(dr.Host))
+				services = append(services, matchedSvcs...)
+				// check for removed subsets
+				prevCfg := push.PrevDestinationRuleByName(proxy, key.Name)
+				if prevCfg == nil {
+					log.Infof("Prev DestinationRule form PrevSidecarScope is nil for %v", key.String())
+					break
+				}
+				prevDr := prevCfg.Spec.(*networking.DestinationRule)
+				// detecting added/updated subsets works, we just need to track deletions
+				log.Infof("Prev DR Subsets: %v\nNew DR subsets: %v", getSubsetNames(prevDr.GetSubsets()), getSubsetNames(dr.GetSubsets()))
+				if !reflect.DeepEqual(prevDr.GetSubsets(), dr.GetSubsets()) {
+					// subsets changed, did something delete?
```

Comment:
what if dr.Host changed?

Reply:
then it's a different destinationrule right? which is applied to a different cluster

Reply:
I am asking the same destinationrule.Host update